File: apogee-backend/apogee/routes/summarize.py
from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse

from apogee.models.request_models import (
    AskRequest,
    SuggestedQuestionsRequest,
    SummaryRequest,
)
from apogee.services.prompt_service import (
    build_answer_prompt,
    build_suggested_questions_prompt,
)
from apogee.services.llm_service import generate_stream, generate_text, LLMError
from apogee.services.summary_service import summarize_text
from apogee.utils.cleaner import clean_text

router = APIRouter()

# Maximum content length accepted (approximately 500 KB of text).
MAX_CONTENT_LENGTH = 500_000


@router.post("/summarize")
def summarize(data: SummaryRequest):
    if len(data.content) > MAX_CONTENT_LENGTH:
        raise HTTPException(
            status_code=413,
            detail=f"Content too large ({len(data.content)} chars). Maximum is {MAX_CONTENT_LENGTH}.",
        )

    return StreamingResponse(
        summarize_text(
            text=data.content,
            title=data.title,
            url=data.url,
            mode=data.mode,
            model=data.model,
        ),
        media_type="text/plain",
    )


@router.post("/ask")
def ask(data: AskRequest):
    if len(data.content) > MAX_CONTENT_LENGTH:
        raise HTTPException(
            status_code=413,
            detail=f"Content too large ({len(data.content)} chars). Maximum is {MAX_CONTENT_LENGTH}.",
        )

    cleaned_content = clean_text(data.content)

    prompt = build_answer_prompt(
        title=data.title,
        url=data.url,
        content=cleaned_content,
        question=data.question,
    )

    def safe_stream():
        try:
            yield from generate_stream(prompt, data.model)
        except LLMError as exc:
            yield f"\n\n[Error: {exc}]"

    return StreamingResponse(
        safe_stream(),
        media_type="text/plain",
    )


def _parse_suggested_questions(text: str) -> list[str]:
    questions = []

    for line in text.splitlines():
        cleaned = line.strip().lstrip("-*•").strip()
        for prefix in ("1.", "2.", "1)", "2)"):
            cleaned = cleaned.removeprefix(prefix).strip()

        if cleaned and cleaned.endswith("?"):
            questions.append(cleaned)

        if len(questions) == 2:
            break

    return questions


@router.post("/suggest-questions")
def suggest_questions(data: SuggestedQuestionsRequest):
    if len(data.summary) > MAX_CONTENT_LENGTH:
        raise HTTPException(
            status_code=413,
            detail=f"Summary too large ({len(data.summary)} chars). Maximum is {MAX_CONTENT_LENGTH}.",
        )

    prompt = build_suggested_questions_prompt(
        title=data.title,
        url=data.url,
        summary=clean_text(data.summary),
    )

    try:
        text = generate_text(prompt, data.model)
    except LLMError as exc:
        raise HTTPException(status_code=502, detail=str(exc)) from exc

    questions = _parse_suggested_questions(text)
    if len(questions) < 2:
        raise HTTPException(
            status_code=502,
            detail="Model did not return two suggested questions.",
        )

    return {"questions": questions}

File: apogee-backend/src/routes/summarize.js
import { Router } from "express";

import {
  AskRequestSchema,
  SuggestedQuestionsRequestSchema,
  SummaryRequestSchema,
} from "../models/requestSchemas.js";
import {
  buildAnswerPrompt,
  buildSuggestedQuestionsPrompt,
} from "../services/promptService.js";
import { generateStream, generateText, LLMError } from "../services/llmService.js";
import { summarizeText } from "../services/summaryService.js";
import { cleanText } from "../utils/cleaner.js";
import { HttpError, validateBody } from "../utils/httpError.js";

const router = Router();

// Maximum content length accepted (approximately 500 KB of text).
const MAX_CONTENT_LENGTH = 500_000;

router.post("/summarize", validateBody(SummaryRequestSchema), async (req, res) => {
  const { content, title, url, mode, model } = req.body;
  if (content.length > MAX_CONTENT_LENGTH) {
    throw new HttpError(
      413,
      `Content too large (${content.length} chars). Maximum is ${MAX_CONTENT_LENGTH}.`,
    );
  }

  res.type("text/plain");
  for await (const token of summarizeText({ text: content, title, url, mode, model })) {
    res.write(token);
  }
  res.end();
});

router.post("/ask", validateBody(AskRequestSchema), async (req, res) => {
  const { content, title, url, question, model } = req.body;
  if (content.length > MAX_CONTENT_LENGTH) {
    throw new HttpError(
      413,
      `Content too large (${content.length} chars). Maximum is ${MAX_CONTENT_LENGTH}.`,
    );
  }

  const prompt = buildAnswerPrompt({
    title,
    url,
    content: cleanText(content),
    question,
  });

  res.type("text/plain");
  try {
    for await (const token of generateStream(prompt, model)) {
      res.write(token);
    }
  } catch (err) {
    if (!(err instanceof LLMError)) throw err;
    res.write(`\n\n[Error: ${err.message}]`);
  }
  res.end();
});

function parseSuggestedQuestions(text) {
  const questions = [];

  for (const rawLine of text.split("\n")) {
    let cleaned = rawLine.trim().replace(/^[-*•]+/, "").trim();
    for (const prefix of ["1.", "2.", "1)", "2)"]) {
      if (cleaned.startsWith(prefix)) {
        cleaned = cleaned.slice(prefix.length).trim();
      }
    }

    if (cleaned && cleaned.endsWith("?")) {
      questions.push(cleaned);
    }

    if (questions.length === 2) break;
  }

  return questions;
}

router.post(
  "/suggest-questions",
  validateBody(SuggestedQuestionsRequestSchema),
  async (req, res) => {
    const { summary, title, url, model } = req.body;
    if (summary.length > MAX_CONTENT_LENGTH) {
      throw new HttpError(
        413,
        `Summary too large (${summary.length} chars). Maximum is ${MAX_CONTENT_LENGTH}.`,
      );
    }

    const prompt = buildSuggestedQuestionsPrompt({
      title,
      url,
      summary: cleanText(summary),
    });

    let text;
    try {
      text = await generateText(prompt, model);
    } catch (err) {
      if (!(err instanceof LLMError)) throw err;
      throw new HttpError(502, err.message);
    }

    const questions = parseSuggestedQuestions(text);
    if (questions.length < 2) {
      throw new HttpError(502, "Model did not return two suggested questions.");
    }

    res.json({ questions });
  },
);

export default router;
export { parseSuggestedQuestions };
